### PR TITLE
+ [ios] LayoutAnimation

### PR DESCRIPTION
**Pull request by doumafang**
Layout Animation Base on CADisplayLink，to create generated frame animation.

### CODE REVIEW

File: ios/sdk/WeexSDK/Sources/Component/WXComponent_internal.h [hunk 1]
```diff
@@ -214,3 +251,11 @@
 - (void)setGradientLayer;
 
 @end
+
+@interface WXLayoutAnimationInfo : NSObject
+@property (nonatomic, strong) id fromValue;
```
Comment (kfeagle) [inline ios/sdk/WeexSDK/Sources/Component/WXComponent_internal.h]: use for CGfloat, maybe assign is better?

File: ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m [hunk 2]
```diff
@@ -112,6 +115,211 @@ - (NSUInteger)_childrenCountForLayout
...
+    [self _suspendLayoutAnimationDisplayLink];
+    
+    if (!_addStyles) {
+        _fromStyles = [NSMutableDictionary dictionaryWithDictionary:self.styles];
+        _addStyles = [NSMutableDictionary dictionaryWithDictionary:styles];
+    }
+    else
+    {
+        [_addStyles addEntriesFromDictionary:styles];
+    }//保证_addStyles是唯一的
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m]: Use English Comments.

File: ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m [hunk 3]
```diff
@@ -112,6 +115,211 @@ - (NSUInteger)_childrenCountForLayout
...
+    [_toStyles addEntriesFromDictionary:_addStyles];
+    
+    _layoutAnimationDuration = _fromStyles[@"transitionDuration"] ? [WXConvert CGFloat:_fromStyles[@"transitionDuration"]] : 0;
+    _layoutAnimationDelay = _fromStyles[@"transitionDelay"] ? [WXConvert CGFloat:_fromStyles[@"transitionDelay"]] : 0;
+    _layoutAnimationTimingFunction = [WXConvert CAMediaTimingFunction:_fromStyles[@"transitionTimingFunction"]];
+    
+    
+    if (_layoutAnimationDuration == 0) {
+        [self _fillCSSNode:styles];
+        return;//如果duration为零直接关闭动画效果
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m]: Use English Comments.

File: ios/sdk/WeexSDK/Sources/Model/WXComponent.m [hunk 4]
```diff
@@ -436,16 +436,35 @@ - (void)_moveToSupercomponent:(WXComponent *)newSupercomponent atIndex:(NSUInteg
 
 - (void)_updateStylesOnComponentThread:(NSDictionary *)styles resetStyles:(NSMutableArray *)resetStyles isUpdateStyles:(BOOL)isUpdateStyles
 {
-    if (isUpdateStyles) {
-        pthread_mutex_lock(&_propertyMutex);
-        [_styles addEntriesFromDictionary:styles];
-        pthread_mutex_unlock(&_propertyMutex);
+    
+    //根据当前的style是否含有transitionProperty属性来判断是否有Layout Animation
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Model/WXComponent.m]: Use English Comments

File: ios/sdk/WeexSDK/Sources/Model/WXComponent.m [hunk 5]
```diff
@@ -436,16 +436,35 @@ - (void)_moveToSupercomponent:(WXComponent *)newSupercomponent atIndex:(NSUInteg
...
+    if (_styles[@"transitionProperty"]) {
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Model/WXComponent.m]: Only transitionProperty?

File: ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m [hunk 6]
```diff
@@ -112,6 +115,211 @@ - (NSUInteger)_childrenCountForLayout
...
+    
+    [self performSelector:@selector(_startLayoutAnimationDisplayLink) withObject:self afterDelay:_layoutAnimationDelay/1000];
+    //    [self _startLayoutAnimationDisplayLink];
+    
+}
+
+
+- (void)_calculateLayoutAnimationProcessingStyle
+{
+    //linear 在做贝塞尔曲线模型
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m]: Use English comments

File: ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m [hunk 7]
```diff
@@ -112,6 +115,211 @@ - (NSUInteger)_childrenCountForLayout
...
+    
+    
+    _widthInfo.isAnimated ? [_fromStyles setObject:@(currentWidth) forKey:@"width"]:0;
+    _heightInfo.isAnimated ? [_fromStyles setObject:@(currentHeight) forKey:@"height"]:0;
+    _leftInfo.isAnimated ? [_fromStyles setObject:@(currentLeft) forKey:@"left"]:0;
+    _rightInfo.isAnimated ? [_fromStyles setObject:@(currentRight) forKey:@"right"]:0;
+    _topInfo.isAnimated ? [_fromStyles setObject:@(currentTop) forKey:@"top"]:0;
+    _bottomInfo.isAnimated ? [_fromStyles setObject:@(currentBottom) forKey:@"bottom"]:0;
+    
+    NSLog(@"%@",_fromStyles);
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m]: Don't use NSLog

File: ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m [hunk 8]
```diff
@@ -409,4 +617,77 @@ static css_dim_t cssNodeMeasure(void *context, float width, css_measure_mode_t w
     return (css_dim_t){resultSize.width, resultSize.height};
 }
 
+//贝塞尔曲线计算
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Layout/WXComponent+Layout.m]: Use English comments

File: ios/sdk/WeexSDK/Sources/Module/WXAnimationLayout.h [hunk 9]
```diff
@@ -1,49 +1,49 @@
...
-@property (nonatomic,strong) WXAnimationLayoutInfo *heightInfo;
-@property (nonatomic,assign) double animationDuration;
-@property (nonatomic,assign) double animationDelay;
-@property (nonatomic,strong) NSDictionary *needUpdateStyles;
-@property (nonatomic, weak) WXSDKInstance *weexInstance;
-
-- (void)layoutForAnimation;
-
-@end
+///*
```
Comment (cxfeng1) [inline ios/sdk/WeexSDK/Sources/Module/WXAnimationLayout.h]: Just delete WXAnimationLayout.h/m

**Comment by kernelj** [comment]
这是增加ios页面跳转的加载动画的吗

**Pull request merged**